ls1intum/Artemis
# `General`: Limit scope of active students line charts by course start and end date

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


We currently provide two line charts displaying the amount of active students per week. Both do not take the course start and end date into account. This has the effect, that for example weeks before the official start of the course are displayed or the active students are shown up to the current week even though the course ended weeks ago. This is changed by this PR. 
### Description

I changed the chart implementation so they are not rendered if the start date is not passed yet and only contain data until the course end date if it is passed.
Therefore I also adapted the server implementation. In case that the end date of a course is already passed, the server only sends the data collected until the end date to the client.
The data sent is also adapted if the visible time span differs from the normal one, e.g. the start date was 1 week ago or the end course life time is shorter than the chart time span.
### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course with an active students distribution

1. Log in to Artemis
2. Navigate to Course Administration
3. Inspect the active students chart for the course. Make sure the values are displayed appropriately. Do the same for the active students chart on course detail page.
4. Edit the course and set the end date e.g. 2 weeks before the current date.
5. Check both charts again. The data should switch accordingly and the data labels should reflect the fact, that old data is displayed (x weeks ago as most right data label on course management and the corresponding ISO week on course detail).
6. Edit the course again. Delete the end date and set the start date on some point in future
7. Check the charts. No charts should be displayed, instead a lettering explains that the course did not start yet.
8. Edit the course again. Set the start date on some point later than 3 weeks before the current date.
9. Check both charts. Now the charts should start from the week of the dedicated start date.
10. Edit the course again. Set the start date more than 16 weeks before the current date. Check the charts. The chart on course management should not be affected by it. The chart on course detail page should only allow going back in time until the start date (by clicking the left arrow)
11. Edit the course again. Set the start date of the course e.g. 2 weeks ago and the end date of the course e.g. 1 week ago. Return to the charts and make sure they change appropriately.
12. Go wild with different combinations of start and end date and make sure no unexpected behavior occurs.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





| File name | Branch coverage (%) | Line coverage (%) |
| -- | -- | -- |
| `CourseRescource.java` | 73 | 93 |
| `CourseService.java` | 75 | 90 |
| `course-detail-line-chart.component.ts` | 79.16 | 94.73 |
| `course-management-overview-statistics.component.ts` | 91.66 | 91.48 |


### Screenshots


### Example screenshot how the course management chart should look after step 4:
![image](https://user-images.githubusercontent.com/80622272/157659914-944df84b-d39e-4d09-9cef-036824296911.png)
For course detail (current date is in CW 10):
![image](https://user-images.githubusercontent.com/80622272/157843401-1702602a-9bb9-404d-aa32-26f8ec0ae2de.png)

### Example after step 6 (similar for course detail):
![image](https://user-images.githubusercontent.com/80622272/164494675-ef70a937-f3bf-4ebb-bb2b-a4ec4660a6f9.png)

### Example after step 8:
![image](https://user-images.githubusercontent.com/80622272/157845724-bf7d60b1-b9a5-4e8a-82c7-2f4a81955450.png)
For course detail:
![image](https://user-images.githubusercontent.com/80622272/157843835-4175d64b-5a6c-49fb-bb3a-8c84d6d05106.png)

<details>
     <summary>Example for course detail after step 10:</summary>

https://user-images.githubusercontent.com/80622272/157844612-a29f249f-b1c3-43f1-bb44-21117589f43c.mp4


</details>

### Example for course management after step 11:
![image](https://user-images.githubusercontent.com/80622272/157844946-b1667270-4db9-495a-a674-62d77d8aa351.png)
For course detail:
![image](https://user-images.githubusercontent.com/80622272/157845013-03972ec4-b7c0-4451-a228-eaa48a2a8eca.png)
